common/poller/FlatMappedPoller.java

```diff
@@ -18,36 +18,43 @@
 
 package com.vaticle.typedb.core.common.poller;
 
-import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
-import com.vaticle.typedb.core.common.iterator.Iterators;
-
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Optional;
 import java.util.function.Function;
 
 public class FlatMappedPoller<T, U> extends AbstractPoller<U> {
 
     private final Poller<T> source;
-    private final Function<T, FunctionalIterator<U>> flatMappingFn;
-    private FunctionalIterator<U> currentIterator;
+    private final Function<T, Poller<U>> flatMappingFn;
+    private final List<Poller<U>> flatMappingPollers;
 
-    public FlatMappedPoller(AbstractPoller<T> poller, Function<T, FunctionalIterator<U>> flatMappingFn) {
+    public FlatMappedPoller(Poller<T> poller, Function<T, Poller<U>> flatMappingFn) {
         this.source = poller;
         this.flatMappingFn = flatMappingFn;
-        currentIterator = Iterators.empty();
+        this.flatMappingPollers = new ArrayList<>();
     }
 
     @Override
     public Optional<U> poll() {
-        mayFetchIterator();
-        if (currentIterator.hasNext()) return Optional.of(currentIterator.next());
-        else return Optional.empty();
+        for (Poller<U> poller : flatMappingPollers) {
+            Optional<U> next = poller.poll();
+            if (next.isPresent()) return next;
+        }
+        Optional<T> fromSource = source.poll();
+        if (fromSource.isPresent()) {
+            Poller<U> newPoller = flatMappingFn.apply(fromSource.get());
+            flatMappingPollers.add(newPoller);
+            return newPoller.poll();
```

Comment:
errr is that enough? if the new poller doesn't have anything you need to loop over the `source` until you've exhausted it, and only if that's all empty and all flatmapped pollers are empty then can you return empty?

Reply:
Actually I think this implementation is enough as you first loop over all previously found `flatMappedPollers` (each derived from an answer from the source). Then we also can't "exhaust" the source as it's a poller.

Reply:
Change made, @flyingsilverfin was right!